beacon-chain/blockchain/core.go

```diff
@@ -107,6 +110,24 @@ func (b *BeaconChain) CanProcessBlock(fetcher powchain.POWBlockFetcher, block *t
 	// Calculate the timestamp validity condition.
 	slotDuration := time.Duration(block.Data().SlotNumber*params.SlotLength) * time.Second
 	validTime := time.Now().After(b.GenesisBlock().Data().Timestamp.Add(slotDuration))
+
+	// Verify state hashes from the block are correct
+	hash, err := hashActiveState(*b.ActiveState())
+	if err != nil {
+		return false, err
+	}
+
+	if !bytes.Equal(block.Data().ActiveStateHash.Sum(nil), hash.Sum(nil)) {
+		return false, fmt.Errorf("Active state hash mismatched, wanted: %v, got: %v", hash.Sum(nil), block.Data().ActiveStateHash.Sum(nil))
+	}
+	hash, err = hashCrystallizedState(*b.CrystallizedState())
+	if err != nil {
+		return false, err
+	}
+	if !bytes.Equal(block.Data().CrystallizedStateHash.Sum(nil), hash.Sum(nil)) {
```

Comment:
This part is likely to change so we can leave this be, but keep in mind that at most one every 100 blocks will contain the crystallized state. Each block's crystallized state can be up to 100MB, depending on the number of validators, and only changes after an epoch transition.

Reply:
That's no longer the case with the latest 2.1 spec. Crystallized state changes every epoch which is `slot_number // EPOCH_LENGTH > epoch_number`  I'll make sure to add a condition for that when 2.1 gets finalized

Reply:
Ah, nice, didn't know that :+1: